tiledb/sm/filesystem/azure.cc

```diff
@@ -649,20 +751,211 @@ Status Azure::touch(const URI& uri) const {
   return Status::Ok();
 }
 
-Status Azure::write(const URI& uri, const void* buffer, uint64_t length) {
+Status Azure::write(
+    const URI& uri, const void* const buffer, const uint64_t length) {
+  if (!uri.is_azure()) {
+    return LOG_STATUS(Status::AzureError(
+        std::string("URI is not an Azure URI: " + uri.to_string())));
+  }
+
+  Buffer* const write_cache_buffer = get_write_cache_buffer(uri.to_string());
+
+  uint64_t nbytes_filled;
+  RETURN_NOT_OK(
+      fill_write_cache(write_cache_buffer, buffer, length, &nbytes_filled));
+
+  if (!use_block_list_upload_) {
+    if (nbytes_filled != length) {
+      std::stringstream errmsg;
+      errmsg << "Direct write failed! " << nbytes_filled
+             << " bytes written to buffer, " << length << " bytes resulted.";
+      return LOG_STATUS(Status::AzureError(errmsg.str()));
+    } else {
+      return Status::Ok();
+    }
+  }
+
+  if (write_cache_buffer->size() == write_cache_max_size_) {
+    RETURN_NOT_OK(flush_write_cache(uri, write_cache_buffer, false));
+  }
+
+  uint64_t new_length = length - nbytes_filled;
+  uint64_t offset = nbytes_filled;
+  while (new_length > 0) {
+    if (new_length >= write_cache_max_size_) {
+      RETURN_NOT_OK(write_blocks(
```

Comment:
can we use `flush_write_cache` here and simplify?

Reply:
We don't want to use `flush_write_cache` because it is specifically used on a `Buffer` object and resets the size.